### ISSUE TITLE
Is it possible to integrate a LM during/after the inference of the non-autoregressive MaskCTC paradigm?

### ISSUE THREAD

**Issue by david-gimeno** [espnet/espnet#5152]
Hello,

I trained an ASR end-to-end system based on MaskCTC. The quality of speech recognition is great and it is amazing how fast the model can decode speech. However, I was wondering if there is a way to integrate the knowledge of a LM during/after the inference process.

Thanks in advance. Best regards,

David.

**Comment by sw005320** [comment]
Currently, MaskCTC in espnet does not have the ability to integrate an LM.
BERT-CTC and BECTRA are one way to do this (https://arxiv.org/pdf/2211.00792.pdf). You can test it in https://github.com/YosukeHiguchi/espnet/tree/bectra

@YosukeHiguchi, we can think of implementing this in the main repository.

**Comment by YosukeHiguchi** [comment]
@sw005320 Yes, I will try to start integrating our BERT-based models into the main repo.

@david-gimeno Please let me know if you have further questions on Mask-CTC.

**Comment by david-gimeno** [comment]
First, the paper you mentioned look so interesting. I will read them for sure! 

Secondly, yes, I would like to ask you something, @YosukeHiguchi:

Why cannot I apply a beam search (where a LM can be integrated in a shallow fusion manner) when using a Mask-CTC model? For instance, when I define a ESPNet model (CTC/Attention architecture) with "ctc_weight=1.0", i.e., a CTC-based ASR model (no Attention-based branch), I can apply a beam search with a LM. Of course, this beam search will deteriorate the real-time factor, but it could be interesting to see how the performance can be improved or just for benchmarking purposes when comparing with models that used a LM.

Thanks in advance,

David.

**Comment by YosukeHiguchi** [comment]
As per your understanding, if you just want to evaluate the performance of the CTC-based encoder without considering the non-autoregressive benefit, it is possible to employ beam search decoding with an LM for a Mask-CTC model.
